doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+  val a = N // id 2
+}
+
+module B::A1 {
+  pure val B::A1::N = 1 // id 5
+  pure val B::A1::a = B::A1::N // id 6
+}
+
+module C::A1 {
+  pure val C::A1::N = 0 // id 7
+  pure val C::A1::a = C::A1::N // id 8
+}
+
+module B {
+  import B::A1.*
+  val b = B::A1::a // id 3
+}
+
+module C {
+  import C::A1.**
+  val c = C::A1::a // id 4
+}
+```
+
+Lookup table before instance flattening:
+```mermaid
+graph BT
+   1["(1) const N: int"]
+   2["(2) val a = N"] --> 1
+   3["(3) val b = A1::a"] --> 2
+   4["(4) val c = A1::a"] --> 2
+```
+
+After instance flattening, there are two independent sub-trees, since both instances are completely independent of each other:
+```mermaid
+graph BT
+  1["(1) const N: int"]
+  2["(2) val a = N"] --> 1
+  3["(3) val b = B::A1::a"] --> 6
+  4["(4) val c = C::A1::a"] --> 8
+  5["(5) pure val B::A1::N = 1"]
+  6["(6) pure val B::A1::a = B::A1::N"] --> 5
+  7["(7) pure val C::A1::N = 0"]
+  8["(8) pure val C::A1::a = C::A1::N"] --> 7
+```
+
+##### Override expressions
+
+In an instance statement like `import A(N=e).*`, `e` might depend on expressions defined in the current module. In this case, we need to ensure that the dependencies of `e` are copied to the new module created for the instance, otherwise `pure val myModule::A::N = e` won't resolve. To accomplish this, we use the actual `Flattener` (described below) to inspect only the expressions in the overrides and find any definitions they might depend on.
+
+### Flattener
+
+The flattener adds missing definitions to a module by recursively fetching the lookup table and copying the referenced definition, then looking into that definition for any other names and doing the same.
+
+When adding a missing definition, if that definitions carries a namespace, we actually need to do a "deep" procedure to add that name to not only the definition name, but also to every other name appearing in it's body. This includes adding the namespace to lambda parameters and nested definitions (even though they cannot be referred), so flattening doesn't create name conflicts with them.
+
+This deep namespace-adding procedure is already done in the current implementation of flattening (see [`addNamespaceToDef`](https://github.com/informalsystems/quint/blob/main/quint/src/flattening.ts#L276C4-L276C4)), with the only difference being that we don't want to generate new ids in this case (we only create new ids in the Instance Flattener).
+
+Although the Instance Flattener requires imports and instances to be flattened, we can't flatten definitions that come from instances before running the instance flattener, since that would copy definitions that would potentially be replaced by new definitions in new modules created by the instance flattener. Therefore, we need to run the flattener twice: first not copying definitions that come from instances (checking the `importedFrom` field once again), and then, after the instance flattener, flattening anything that remains to be flattened.
```

Comment:
I don't follow here: you say we cannot flatten before instance-flattening, but why can't we instance-flatten and then flatten?

What happens if we have a instance->import->instance->import->... chain? Does this degenerate into a recursive procedure?

Reply:
> but why can't we instance-flatten and then flatten?

Because the Instance Flattener requires imports and instances to be flattened.

> Does this degenerate into a recursive procedure?

It doesn't. One round of flattener -> instance flattener -> flattener is enough. That's because the first round of flattening will get rid of all imports and exports, leaving us with only instances. The instance flattener will then replace the instances with imports, leaving us with only imports. The last round of flattener gets rid of the remaining imports.

We could maybe avoid the last flattening round if, in the instance flattener, instead of generating an import statement, we did some inner flattening as well. But I think that would only complicate the code and result in no benefit. With the current approach, there is a clearer (or at least less unclear) separation of concerns between the flattener and the instance flattener.